File: ios/TurboModules/LocationModuleLegacy.m
/**
 * 🐌 LEGACY MODULE - Location Services (iOS)
 * 
 * Traditional React Native bridge implementation with CoreLocation
 * Goes through bridge queue with JSON serialization (slower)
 */

#import "LocationModuleLegacy.h"
#import <CoreLocation/CoreLocation.h>

@interface LocationModuleLegacy () <CLLocationManagerDelegate>
@property (nonatomic, strong) CLLocationManager *locationManager;
@property (nonatomic, strong) CLLocation *lastKnownLocation;
@end

@implementation LocationModuleLegacy

RCT_EXPORT_MODULE()

- (instancetype)init {
    self = [super init];
    if (self) {
        self.locationManager = [[CLLocationManager alloc] init];
        self.locationManager.delegate = self;
        self.locationManager.desiredAccuracy = kCLLocationAccuracyBest;
    }
    return self;
}

RCT_EXPORT_METHOD(requestLocationPermission:(RCTPromiseResolveBlock)resolve
                                     reject:(RCTPromiseRejectBlock)reject) {
    // Simulate bridge overhead with small delay
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.005 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        @try {
            CLAuthorizationStatus status = [CLLocationManager authorizationStatus];
            
            switch (status) {
                case kCLAuthorizationStatusAuthorizedWhenInUse:
                case kCLAuthorizationStatusAuthorizedAlways:
                    resolve(@YES);
                    break;
                case kCLAuthorizationStatusNotDetermined:
                    [self.locationManager requestWhenInUseAuthorization];
                    resolve(@NO);
                    break;
                default:
                    resolve(@NO);
                    break;
            }
        } @catch (NSException *exception) {
            reject(@"PERMISSION_ERROR", @"Failed to check location permission", nil);
        }
    });
}

RCT_EXPORT_METHOD(getCurrentLocation:(RCTPromiseResolveBlock)resolve
                              reject:(RCTPromiseRejectBlock)reject) {
    // Simulate bridge overhead with small delay
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.010 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        @try {
            // Check authorization
            CLAuthorizationStatus status = [CLLocationManager authorizationStatus];
            if (status != kCLAuthorizationStatusAuthorizedWhenInUse && 
                status != kCLAuthorizationStatusAuthorizedAlways) {
                reject(@"PERMISSION_DENIED", @"Location permission not granted", nil);
                return;
            }

            // Get location (same logic as TurboModule but with bridge overhead)
            CLLocation *location = self.locationManager.location ?: self.lastKnownLocation;
            
            if (location) {
                NSDictionary *locationData = @{
                    @"latitude": @(location.coordinate.latitude),
                    @"longitude": @(location.coordinate.longitude),
                    @"accuracy": @(location.horizontalAccuracy),
                    @"altitude": @(location.altitude),
                    @"speed": @(location.speed),
                    @"timestamp": @([location.timestamp timeIntervalSince1970] * 1000)
                };
                resolve(locationData);
            } else {
                // Generate mock location for demo
                double lat = 37.7749 + ((double)arc4random() / UINT32_MAX - 0.5) * 0.01;
                double lng = -122.4194 + ((double)arc4random() / UINT32_MAX - 0.5) * 0.01;
                double accuracy = 15.0 + ((double)arc4random() / UINT32_MAX) * 25.0; // Slightly less accurate
                
                NSDictionary *mockLocation = @{
                    @"latitude": @(lat),
                    @"longitude": @(lng),
                    @"accuracy": @(accuracy),
                    @"altitude": @(50.0 + ((double)arc4random() / UINT32_MAX) * 100.0),
                    @"speed": @(0.0),
                    @"timestamp": @([[NSDate date] timeIntervalSince1970] * 1000)
                };
                resolve(mockLocation);
            }
        } @catch (NSException *exception) {
            reject(@"LOCATION_ERROR", @"Failed to get current location", nil);
        }
    });
}

#pragma mark - CLLocationManagerDelegate

- (void)locationManager:(CLLocationManager *)manager 
     didUpdateLocations:(NSArray<CLLocation *> *)locations {
    self.lastKnownLocation = locations.lastObject;
}

- (void)locationManager:(CLLocationManager *)manager 
       didFailWithError:(NSError *)error {
    NSLog(@"Location error: %@", error.localizedDescription);
}

@end

File: android/app/src/main/java/com/turbomodules/legacy/LocationModuleLegacy.kt
package com.turbomodules.legacy

import android.Manifest
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import androidx.core.content.ContextCompat
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableNativeMap

/**
 * 🐌 LEGACY MODULE - Location Services (Android)
 * 
 * Traditional React Native bridge implementation
 * Goes through bridge queue with JSON serialization (slower)
 */
class LocationModuleLegacy(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    override fun getName() = "LocationModuleLegacy"

    private val locationManager: LocationManager by lazy {
        reactApplicationContext.getSystemService(android.content.Context.LOCATION_SERVICE) as LocationManager
    }

    @ReactMethod
    fun requestLocationPermission(promise: Promise) {
        try {
            // Add some artificial delay to simulate bridge overhead
            Thread.sleep(5) // 5ms delay
            
            val hasPermission = ContextCompat.checkSelfPermission(
                reactApplicationContext,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED

            promise.resolve(hasPermission)
        } catch (e: Exception) {
            promise.reject("PERMISSION_ERROR", "Failed to check location permission", e)
        }
    }

    @ReactMethod
    fun getCurrentLocation(promise: Promise) {
        try {
            // Add some artificial delay to simulate bridge overhead
            Thread.sleep(10) // 10ms delay
            
            // Check permission
            if (ContextCompat.checkSelfPermission(
                    reactApplicationContext,
                    Manifest.permission.ACCESS_FINE_LOCATION
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                promise.reject("PERMISSION_DENIED", "Location permission not granted")
                return
            }

            // Get location (same logic as TurboModule but with bridge overhead)
            val location = getLastKnownLocation()
            
            if (location != null) {
                val locationData = WritableNativeMap().apply {
                    putDouble("latitude", location.latitude)
                    putDouble("longitude", location.longitude)
                    putDouble("accuracy", location.accuracy.toDouble())
                    putDouble("altitude", location.altitude)
                    putDouble("speed", location.speed.toDouble())
                    putDouble("timestamp", location.time.toDouble())
                }
                promise.resolve(locationData)
            } else {
                // Generate mock location for demo
                val mockLocation = WritableNativeMap().apply {
                    putDouble("latitude", 37.7749 + (Math.random() - 0.5) * 0.01)
                    putDouble("longitude", -122.4194 + (Math.random() - 0.5) * 0.01)
                    putDouble("accuracy", 15.0 + Math.random() * 25) // Slightly less accurate
                    putDouble("altitude", 50.0 + Math.random() * 100)
                    putDouble("speed", 0.0)
                    putDouble("timestamp", System.currentTimeMillis().toDouble())
                }
                promise.resolve(mockLocation)
            }
        } catch (e: Exception) {
            promise.reject("LOCATION_ERROR", "Failed to get current location: ${e.message}", e)
        }
    }

    private fun getLastKnownLocation(): Location? {
        return try {
            val providers = listOf(
                LocationManager.GPS_PROVIDER,
                LocationManager.NETWORK_PROVIDER,
                LocationManager.PASSIVE_PROVIDER
            )

            var bestLocation: Location? = null
            
            for (provider in providers) {
                if (locationManager.isProviderEnabled(provider)) {
                    val location = locationManager.getLastKnownLocation(provider)
                    if (location != null && (bestLocation == null || location.accuracy < bestLocation.accuracy)) {
                        bestLocation = location
                    }
                }
            }
            
            bestLocation
        } catch (e: SecurityException) {
            null
        }
    }
}